File: tests/test_tokens.py
from __future__ import annotations

import os
import stat

from lcpush import tokens
from lcpush.paths import token_file


def test_env_var_precedence(monkeypatch):
    monkeypatch.setenv("GITHUB_TOKEN", "from-github-token")
    monkeypatch.setenv("LCPUSH_GITHUB_TOKEN", "from-lcpush-token")
    found = tokens.resolve()
    assert found.value == "from-lcpush-token"
    assert found.source == "$LCPUSH_GITHUB_TOKEN"


def test_env_fallback_to_github_token(monkeypatch):
    monkeypatch.setenv("GITHUB_TOKEN", "gh-token")
    assert tokens.from_env().source == "$GITHUB_TOKEN"


def test_blank_env_is_ignored(monkeypatch):
    monkeypatch.setenv("LCPUSH_GITHUB_TOKEN", "   ")
    assert tokens.from_env() is None


def test_file_fallback_is_0600(monkeypatch):
    monkeypatch.setattr(tokens, "_keyring", lambda: None)
    where = tokens.store("secret-value")
    assert where == str(token_file())
    assert stat.S_IMODE(os.stat(token_file()).st_mode) == 0o600
    assert tokens.from_file().value == "secret-value"


def test_keyring_is_preferred(monkeypatch):
    store: dict[tuple[str, str], str] = {}

    class FakeKeyring:
        @staticmethod
        def set_password(service, account, value):
            store[(service, account)] = value

        @staticmethod
        def get_password(service, account):
            return store.get((service, account))

        @staticmethod
        def delete_password(service, account):
            store.pop((service, account), None)

    monkeypatch.setattr(tokens, "_keyring", lambda: FakeKeyring)
    assert tokens.store("kr-token") == "keyring"
    assert tokens.from_keyring().value == "kr-token"
    assert tokens.clear() == ["keyring"]
    assert tokens.from_keyring() is None


def test_keyring_failure_falls_back_to_file(monkeypatch):
    class BrokenKeyring:
        @staticmethod
        def set_password(*_args):
            raise RuntimeError("no backend")

        @staticmethod
        def get_password(*_args):
            raise RuntimeError("no backend")

    monkeypatch.setattr(tokens, "_keyring", lambda: BrokenKeyring)
    assert tokens.store("fallback") == str(token_file())
    assert tokens.from_keyring() is None


def test_gh_cli_used_when_present(monkeypatch):
    monkeypatch.setattr(tokens, "_keyring", lambda: None)
    monkeypatch.setattr(tokens.shutil, "which", lambda _name: "/usr/bin/gh")

    class Result:
        returncode = 0
        stdout = "gho_from_cli\n"

    monkeypatch.setattr(tokens.subprocess, "run", lambda *a, **k: Result())
    found = tokens.resolve()
    assert found.value == "gho_from_cli"
    assert found.source == "gh auth token"


def test_gh_cli_absent(monkeypatch):
    monkeypatch.setattr(tokens.shutil, "which", lambda _name: None)
    assert tokens.from_gh_cli() is None


def test_resolve_returns_none_when_nothing_stored(monkeypatch):
    monkeypatch.setattr(tokens, "_keyring", lambda: None)
    monkeypatch.setattr(tokens.shutil, "which", lambda _name: None)
    assert tokens.resolve() is None


def test_clear_removes_the_file(monkeypatch):
    monkeypatch.setattr(tokens, "_keyring", lambda: None)
    tokens.store("bye")
    assert tokens.clear() == [str(token_file())]
    assert not token_file().exists()


def test_redact():
    assert tokens.redact("url ghp_abcdefgh here", "ghp_abcdefgh") == "url **** here"
    assert tokens.redact("nothing", None) == "nothing"
    assert tokens.redact("short", "abc") == "short"

File: tests/test_tokens.cpp
#include <catch2/catch_test_macros.hpp>

#include <sys/stat.h>

#include "helpers/fake_subprocess.hpp"
#include "helpers/fixtures.hpp"
#include "lcpush/platform/paths.hpp"
#include "lcpush/github/tokens.hpp"

using namespace lcpush;
using lcpush::testing::EnvVar;
using lcpush::testing::FakeKeyring;
using lcpush::testing::FakeSubprocess;
using lcpush::testing::IsolatedDirs;

namespace {

// Most token tests want "no keyring, no gh CLI" as the baseline.
struct NoBackends {
    keyring::KeyringOverride no_keyring{nullptr};
    FakeSubprocess proc;
    util::SubprocessOverride subprocess_override{&proc};
};

}  // namespace

TEST_CASE("env var precedence") {
    IsolatedDirs dirs;
    EnvVar github("GITHUB_TOKEN", "from-github-token");
    EnvVar lcpush_var("LCPUSH_GITHUB_TOKEN", "from-lcpush-token");
    auto found = tokens::resolve();
    REQUIRE(found.has_value());
    CHECK(found->value == "from-lcpush-token");
    CHECK(found->source == "$LCPUSH_GITHUB_TOKEN");
}

TEST_CASE("env fallback to github token") {
    IsolatedDirs dirs;
    EnvVar github("GITHUB_TOKEN", "gh-token");
    auto found = tokens::from_env();
    REQUIRE(found.has_value());
    CHECK(found->source == "$GITHUB_TOKEN");
}

TEST_CASE("blank env is ignored") {
    IsolatedDirs dirs;
    EnvVar blank("LCPUSH_GITHUB_TOKEN", "   ");
    CHECK_FALSE(tokens::from_env().has_value());
}

TEST_CASE("file fallback is 0600") {
    IsolatedDirs dirs;
    NoBackends backends;
    auto where = tokens::store("secret-value");
    CHECK(where == paths::token_file().string());
    struct stat info{};
    REQUIRE(::stat(paths::token_file().c_str(), &info) == 0);
    CHECK((info.st_mode & 0777) == 0600);
    auto found = tokens::from_file();
    REQUIRE(found.has_value());
    CHECK(found->value == "secret-value");
}

TEST_CASE("keyring is preferred") {
    IsolatedDirs dirs;
    FakeKeyring fake;
    keyring::KeyringOverride keyring_override(&fake);
    CHECK(tokens::store("kr-token") == "keyring");
    auto found = tokens::from_keyring();
    REQUIRE(found.has_value());
    CHECK(found->value == "kr-token");
    CHECK(tokens::clear() == std::vector<std::string>{"keyring"});
    CHECK_FALSE(tokens::from_keyring().has_value());
}

TEST_CASE("keyring failure falls back to file") {
    IsolatedDirs dirs;
    FakeKeyring broken;
    broken.broken = true;
    keyring::KeyringOverride keyring_override(&broken);
    CHECK(tokens::store("fallback") == paths::token_file().string());
    CHECK_FALSE(tokens::from_keyring().has_value());
}

TEST_CASE("gh cli used when present") {
    IsolatedDirs dirs;
    keyring::KeyringOverride no_keyring(nullptr);
    FakeSubprocess proc;
    proc.on_which = [](const std::string&) { return std::optional<std::string>{"/usr/bin/gh"}; };
    proc.on_run_capture = [](const std::vector<std::string>&) {
        return util::RunResult{0, "gho_from_cli\n", true};
    };
    util::SubprocessOverride subprocess_override(&proc);
    auto found = tokens::resolve();
    REQUIRE(found.has_value());
    CHECK(found->value == "gho_from_cli");
    CHECK(found->source == "gh auth token");
}

TEST_CASE("gh cli absent") {
    IsolatedDirs dirs;
    NoBackends backends;
    CHECK_FALSE(tokens::from_gh_cli().has_value());
}

TEST_CASE("resolve returns none when nothing stored") {
    IsolatedDirs dirs;
    NoBackends backends;
    CHECK_FALSE(tokens::resolve().has_value());
}

TEST_CASE("clear removes the file") {
    IsolatedDirs dirs;
    NoBackends backends;
    tokens::store("bye");
    CHECK(tokens::clear() == std::vector<std::string>{paths::token_file().string()});
    CHECK_FALSE(std::filesystem::exists(paths::token_file()));
}

TEST_CASE("redact") {
    CHECK(tokens::redact("url ghp_abcdefgh here", std::string("ghp_abcdefgh")) ==
          "url **** here");
    CHECK(tokens::redact("nothing", std::nullopt) == "nothing");
    CHECK(tokens::redact("short", std::string("abc")) == "short");
}
